### PR TITLE
Trusty fixes

### PR DESCRIPTION
**Pull request by arve-android**
Trusty spd updates. Restore trusty setup code and add some missing patches.

### CODE REVIEW

File: services/spd/trusty/trusty.c [hunk 1]
```diff
@@ -327,22 +327,22 @@ static int32_t trusty_init(void)
 	return 0;
 }
 
-static void trusty_cpu_suspend(void)
+static void trusty_cpu_suspend(uint32_t off)
 {
 	struct args ret;
 
-	ret = trusty_context_switch(NON_SECURE, SMC_FC_CPU_SUSPEND, 0, 0, 0);
+	ret = trusty_context_switch(NON_SECURE, SMC_FC_CPU_SUSPEND, off, 0, 0);
```
Comment (vwadekar) [inline services/spd/trusty/trusty.c]: How does Trusty handle this code?

File: services/spd/trusty/trusty.c [hunk 2]
```diff
@@ -363,18 +363,18 @@ static void trusty_cpu_on_finish_handler(uint64_t unused)
...
 {
-	trusty_cpu_suspend();
+	trusty_cpu_suspend(0);
 }
 
 static void trusty_cpu_suspend_finish_handler(uint64_t unused)
 {
-	trusty_cpu_resume();
+	trusty_cpu_resume(0);
 }
```
Comment (vwadekar) [inline services/spd/trusty/trusty.c]: In general does it make sense to pass the actual suspend state to Trusty? I suppose TAs can definitely use the System Suspend state. Im not sure how that case would be handled with just two status values

File: services/spd/trusty/trusty.c [hunk 3]
```diff
@@ -448,6 +448,31 @@ static int32_t trusty_setup(void)
...
+			spsr |= MODE32_svc << MODE32_SHIFT;
+		}
+		if (spsr != ns_ep_info->spsr) {
+			NOTICE("Trusty: Switch bl33 from EL2 to EL1 (spsr 0x%x -> 0x%x)\n",
+			       ns_ep_info->spsr, spsr);
+			ns_ep_info->spsr = spsr;
+		}
+	}
+
 	return 0;
```
Comment (vwadekar) [inline services/spd/trusty/trusty.c]: Doesnt this break the case when we want to run in EL2 mode? e.g. with a hypervisor?

File: services/spd/trusty/generic-arm64-smcall.c [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+		for (i = 0; i < s->l; i++) {
+			putchar(s->linebuf[i]);
+		}
+		if (ch != '\n') {
+			printf(" <...>\n");
+		}
+		s->l = 0;
+	}
+}
+
```
Comment (vwadekar) [inline services/spd/trusty/generic-arm64-smcall.c]: Why do we need Trusty to send an SMC for printing to the console?

File: services/spd/trusty/generic-arm64-smcall.c [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+static uint64_t trusty_get_reg_base(uint32_t reg)
+{
+	switch (reg) {
+	case 0:
+		return PLAT_ARM_GICD_BASE;
+
+	case 1:
+		return PLAT_ARM_GICC_BASE;
+
```
Comment (vwadekar) [inline services/spd/trusty/generic-arm64-smcall.c]: Doesnt the Trusty platform code, running in S-EL1, already provide these settings?

File: services/spd/trusty/trusty.c [hunk 6]
```diff
@@ -385,11 +385,23 @@ static const spd_pm_ops_t trusty_pm = {
 	.svc_suspend_finish = trusty_cpu_suspend_finish_handler,
 };
 
+void plat_trusty_set_boot_args(aapcs64_params_t *args);
+
+#ifdef TSP_SEC_MEM_SIZE
+#pragma weak plat_trusty_set_boot_args
+void plat_trusty_set_boot_args(aapcs64_params_t *args)
+{
+	args->arg0 = TSP_SEC_MEM_SIZE;
```
Comment (vwadekar) [inline services/spd/trusty/trusty.c]: Why did you remove arg1 and arg2 from the equation?

File: services/spd/trusty/trusty.c [hunk 7]
```diff
@@ -398,17 +410,29 @@ static int32_t trusty_setup(void)
...
-	SET_PARAM_HEAD(ep_info, PARAM_EP, VERSION_1, SECURE | EP_ST_ENABLE);
-	ep_info->spsr = SPSR_64(MODE_EL1, MODE_SP_ELX, DISABLE_ALL_EXCEPTIONS);
+	instr = *(uint32_t *)ep_info->pc;
 
-	/*
-	 * arg0 = TZDRAM aperture available for BL32
-	 * arg1 = BL32 boot params
-	 * arg2 = BL32 boot params length
-	 */
-	ep_info->args.arg1 = ep_info->args.arg2;
```
Comment (vwadekar) [inline services/spd/trusty/trusty.c]: Are these parameters no longer required by Trusty?

Comment (arve-android) [reply services/spd/trusty/trusty.c]: These are platform specific. If you need them on your platform, you can set them in plat_trusty_set_boot_args.

File: services/spd/trusty/trusty.c (hunk 6 above)
Comment (arve-android) [reply services/spd/trusty/trusty.c]: They are platform specific, I don't know where to get them from in generic code.

File: services/spd/trusty/trusty.c (hunk 2 above)
Comment (arve-android) [reply services/spd/trusty/trusty.c]: Do we have access to the system suspend state here? There is a separate psci call for system suspend, but that information seems to be lost before these hooks are called.

File: services/spd/trusty/trusty.c (hunk 1 above)
Comment (arve-android) [reply services/spd/trusty/trusty.c]: It is used to disable fiqs when a cpu is unplugged but leave them on when a cpu enter idle.

File: services/spd/trusty/generic-arm64-smcall.c (hunk 4 above)
Comment (arve-android) [reply services/spd/trusty/generic-arm64-smcall.c]: The generic-arm64 target has no platform specific code.

File: services/spd/trusty/generic-arm64-smcall.c (hunk 5 above)
Comment (arve-android) [reply services/spd/trusty/generic-arm64-smcall.c]: This is only for the generic-arm64 target. You don't need to enable this if you have a platform specific trusty image.

File: services/spd/trusty/generic-arm64-smcall.c (hunk 4 above)
Comment (vwadekar) [reply services/spd/trusty/generic-arm64-smcall.c]: What is restricting you from adding that platform code? In general, issuing an SMC for printing to console does not look like an elegant solution.

File: services/spd/trusty/trusty.c (hunk 6 above)
Comment (vwadekar) [reply services/spd/trusty/trusty.c]: In some implementations the previous level bootloader passes this information to the Trusted OS - memory aperture, uart port, etc. I think we should have that model and then sanity check the boot params in the SPD, instead of coming with another way of passing boot params.

File: services/spd/trusty/trusty.c (hunk 7 above)
Comment (vwadekar) [reply services/spd/trusty/trusty.c]: But now you removed the previous code. Shouldnt we be making changes in such a way that current platforms do not break?

File: services/spd/trusty/trusty.c (hunk 2 above)
Comment (vwadekar) [reply services/spd/trusty/trusty.c]: I see that system_suspend eventually calls into this trusty_cpu_suspend() with the max suspend level. So trusty SPD should be able to differentiate easily, AFAICS.

File: services/spd/trusty/trusty.c (hunk 7 above)
Comment (arve-android) [reply services/spd/trusty/trusty.c]: Can you add the hooks to your platform code? This patch restores the part of the original trusty patches we need to boot on other platforms. Those platforms stopped working after syncing your changes.  I don't know why you decided to remove the mechanism we had for platform specific initialization and add your init code to the generic trusty code instead. Since I don't know which platform(s) rely on this, and since I cannot test them, I did not touch them.

File: services/spd/trusty/generic-arm64-smcall.c (hunk 4 above)
Comment (arve-android) [reply services/spd/trusty/generic-arm64-smcall.c]: This code is only used for the generic-arm64 target. Adding platform specific code to that target would make it non-generic.

File: services/spd/trusty/trusty.c [hunk 8]
```diff
@@ -299,6 +295,7 @@ static int32_t trusty_init(void)
 	ep_info = bl31_plat_get_next_image_ep_info(SECURE);
 	assert(ep_info);
 
+	fpregs_context_save(get_fpregs_ctx(cm_get_context(NON_SECURE)));
```
[the remark this answers is not in the document]
Comment (arve-android) [reply services/spd/trusty/trusty.c]: CTX_INCLUDE_FPREGS is enabled in the make file and required for trusty.

[the remark this answers is not in the document]
Comment (davidcunado-arm) [reply services/spd/trusty/trusty.c]: Apologies missed that bit! And I see that you've also remove the guards around the existing calls to fpregs_context_restore/save.

**Comment by davidcunado-arm** [comment]
@vwadekar 
Please can you progress this with @arve-android ? I won't look to start the merge process until I get your nod - please ping me when ready?

**Comment by davidcunado-arm** [comment]
jenkins: test this please

**Comment by vwadekar** [comment]
Sorry I couldn't get to this PR earlier. I will try to spend some time
today.

**Comment by vwadekar** [comment]
@arve-android

Can you please respond to the comments that are still open?

File: services/spd/trusty/trusty.c (hunk 2 above)
Comment (arve-android) [reply services/spd/trusty/trusty.c]: We can add more values later if we need to know about system suspend. For now I don't want to add a hack to try to guess if we are entering system suspend based on the platform specific power level.

File: services/spd/trusty/trusty.c (hunk 6 above)
Comment (arve-android) [reply services/spd/trusty/trusty.c]: see comment below

**Comment by arve-android** [comment]
@vwadekar is there a specific comment you want me to respond to? I replied to the two comments that did not have a direct reply from me, but it is not clear to me why you would be waiting on a reply to those specific comments.

**Comment by davidcunado-arm** [comment]
@arve-android @vwadekar 
Check failures include build errors on following configs:
- PLAT=tegra TARGET_SOC=t186 SPD=trusty
- PLAT=tegra TARGET_SOC=t210 SPD=trusty

Build error (which I've not looked into):

./build/tegra/t186/debug/bl31/trusty.o: In function `trusty_setup':
/work/workspace/trusted-firmware-build/BUILD_CONFIG/tegra-t186-trusty/MODE/Debug/label/platforms/services/spd/trusty/trusty.c:434: undefined reference to `plat_trusty_set_boot_args'
/work/workspace/trusted-firmware-build/BUILD_CONFIG/tegra-t186-trusty/MODE/Debug/label/platforms/services/spd/trusty/trusty.c:434:(.text.trusty_setup+0xa4): relocation truncated to fit: R_AARCH64_CALL26 against undefined symbol `plat_trusty_set_boot_args'
make: *** [build/tegra/t186/debug/bl31/bl31.elf] Error 1
Build failed with code 2

**Comment by vwadekar** [comment]
@arve-android

Since you are already touching the trusty handlers,  makes sense  to add System suspend as well. We can start another PR if you want.

But I think you should definitely add documentation for the newly added platform handler, as part of this PR, so that it gets easy to port.

File: services/spd/trusty/trusty.c (hunk 7 above)
Comment (arve-android) [reply services/spd/trusty/trusty.c]: I added a patch to add a plat_trusty_set_boot_args hook to the tegra platform. Please test it as I don't have any hardware to test it on.

File: services/spd/trusty/trusty.c [hunk 9]
```diff
@@ -398,17 +406,29 @@ static int32_t trusty_setup(void)
...
+		INFO("trusty: Found 32 bit image\n");
+		aarch32 = 1;
+	} else if (instr >> 8 == 0xd53810 || instr >> 16 == 0x9400) {
+		INFO("trusty: Found 64 bit image\n");
+	} else {
+		NOTICE("trusty: Found unknown image, 0x%x\n", instr);
+	}
+
+	SET_PARAM_HEAD(ep_info, PARAM_EP, VERSION_1, SECURE | EP_ST_ENABLE);
+	if (!aarch32)
```
Comment (davidcunado-arm) [inline services/spd/trusty/trusty.c]: misra_violation: [Required] MISRA C-2012 Rule 10.1 violation: The operand aarch32 has essential type signed 32-bit int, but should be Essentially Boolean.

Change to:
if (aarch32 == 0)

File: services/spd/trusty/trusty.c [hunk 10]
```diff
@@ -398,17 +406,29 @@ static int32_t trusty_setup(void)
...
+	instr = *(uint32_t *)ep_info->pc;
 
-	/*
-	 * arg0 = TZDRAM aperture available for BL32
-	 * arg1 = BL32 boot params
-	 * arg2 = BL32 boot params length
-	 */
-	ep_info->args.arg1 = ep_info->args.arg2;
-	ep_info->args.arg2 = TRUSTY_PARAMS_LEN_BYTES;
+	if (instr >> 24 == 0xea) {
```
Comment (davidcunado-arm) [inline services/spd/trusty/trusty.c]: 1. misra_violation: [Required] MISRA C-2012 Rule 10.4 violation: The operands instr>>24 and 0xea have essential type categories unsigned 32-bit int and signed 16-bit int, which do not match.

Change to:
if (instr >> 24 == 0xeaU)

File: services/spd/trusty/trusty.c [hunk 11]
```diff
@@ -425,6 +445,31 @@ static int32_t trusty_setup(void)
 	if (ret)
 		ERROR("trusty: failed to register fiq handler, ret = %d\n", ret);
 
+	if (aarch32) {
```
Comment (davidcunado-arm) [inline services/spd/trusty/trusty.c]: misra_violation: [Required] MISRA C-2012 Rule 14.4 violation: Condition aarch32 is not essentially boolean.

Change to:
if (aarch32 != 0)

File: services/spd/trusty/trusty.c [hunk 12]
```diff
@@ -398,17 +406,29 @@ static int32_t trusty_setup(void)
...
+	if (!aarch32)
+		ep_info->spsr = SPSR_64(MODE_EL1, MODE_SP_ELX,
+					DISABLE_ALL_EXCEPTIONS);
+	else
+		ep_info->spsr = SPSR_MODE32(MODE32_svc, SPSR_T_ARM,
+					    SPSR_E_LITTLE,
+					    DAIF_FIQ_BIT |
+					    DAIF_IRQ_BIT |
+					    DAIF_ABT_BIT);
+	memset(&ep_info->args, 0, sizeof(ep_info->args));
```
Comment (davidcunado-arm) [inline services/spd/trusty/trusty.c]: misra_violation: [Required] MISRA C-2012 Rule 17.7 violation: Return value of non-void function memset() is not used.

The function trusty_setup returns uint_32, so could just return the error code from memset?

File: services/spd/trusty/generic-arm64-smcall.c [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+static uint64_t trusty_generic_platform_smc(uint32_t smc_fid,
+			 uint64_t x1,
+			 uint64_t x2,
+			 uint64_t x3,
+			 uint64_t x4,
+			 void *cookie,
+			 void *handle,
+			 uint64_t flags)
+{
+	switch(smc_fid) {
```
Comment (davidcunado-arm) [inline services/spd/trusty/generic-arm64-smcall.c]: Style check failure:
services/spd/trusty/generic-arm64-smcall.c:69: ERROR: space required before the open parenthesis '('
+	switch(smc_fid) {

File: services/spd/trusty/trusty.c (hunk 9 above)
Comment (arve-android) [reply services/spd/trusty/trusty.c]: changed type

File: services/spd/trusty/trusty.c (hunk 11 above)
Comment (arve-android) [reply services/spd/trusty/trusty.c]: changed type

File: services/spd/trusty/trusty.c (hunk 12 above)
Comment (arve-android) [reply services/spd/trusty/trusty.c]: memset does not return an error it returns the first argument. I'm not sure how this error can be corrected in a reasonable way

Comment (danh-arm) [reply services/spd/trusty/trusty.c]: Just cast the return value to void:

`(void)memset(&ep_info->args, 0, sizeof(ep_info->args));`

File: services/spd/trusty/trusty.c [hunk 14]
```diff
@@ -398,17 +407,29 @@ static int32_t trusty_setup(void)
...
-	 * arg0 = TZDRAM aperture available for BL32
-	 * arg1 = BL32 boot params
-	 * arg2 = BL32 boot params length
-	 */
-	ep_info->args.arg1 = ep_info->args.arg2;
-	ep_info->args.arg2 = TRUSTY_PARAMS_LEN_BYTES;
+	if (instr >> 24 == 0xeaU) {
+		INFO("trusty: Found 32 bit image\n");
+		aarch32 = true;
+	} else if (instr >> 8 == 0xd53810 || instr >> 16 == 0x9400) {
```
Comment (davidcunado-arm) [inline services/spd/trusty/trusty.c]: @arve-android 
Similar to change to line 412, there also need to have U suffix to the RH operands of ==.

(Apologies, not sure why I didn't comment this before)

**Comment by arve-android** [comment]
@davidcunado-arm 
I'm not in a hurry to merge this time wise, but I would prefer to get it into the release. The last release broke everyone using the patches we provided to add trusty support. It would be nice if we can tell them to get the next release instead.

**Comment by davidcunado-arm** [comment]
@vwadekar 
Happy for me to merge this?

**Comment by vwadekar** [comment]
I would still like the newly added platform handler to be documented before
we merge. Arve?

**Comment by davidcunado-arm** [comment]
@vwadekar @arve-android 
Ok, we're doing a RC1 probably on MON / TUE - can you get this missing piece done on MON 5-MAR?

**Comment by vwadekar** [comment]
Thanks David. I am good with the PR once we add some documentation.

Arve, what do you think? Would you be able to push an update before the cut
off date?

**Comment by arve-android** [comment]
@vwadekar 
The platform handler is not new (I don't know why you removed it in your initial upload), but it has been added to the documentation now.

**Comment by vwadekar** [comment]
Looks good. David we can go ahead and merge this one.

**Comment by davidcunado-arm** [comment]
jenkins: test this please

**Comment by davidcunado-arm** [comment]
checks - passed (false positives in the static checks).
